### ISSUE TITLE
Equivalent of PS2?

### ISSUE THREAD

**Issue by aureooms** [fish-shell/fish-shell#3382]
Is there an equivalent of `PS2` in fish? For example, for bash I have
```bash
PS2="\[${yellow}\]→ \[${reset}\]";
```
where `${yellow}` and `${reset}` are equivalent to `set_color -o yellow` and `set_color normal`.

**Comment by faho** [comment]
For those not familiar, here's the relevant excerpt from the bash man page:

> PS2    The value of this parameter is expanded as with PS1 and used as the secondary prompt string.  The default is ``> ''.

As far as I know, this pretty much means it's used for everything after the first line.

So it'll look like

```fish
> if true; then
$PS2
```

(where $PS2 is the value of PS2 and ">" is the value of PS1 - the normal prompt).

----

Since we don't print a prompt for everything after the first line, there is no such thing in fish.

Any particular reason why you'd want it? Personally I'm okay with the other lines just hanging there.

**Comment by aureooms** [comment]
It would make the distinction between a slow executing command and an uncomplete command clearer than just the position of the cursor.

**Comment by floam** [comment]
Fish doesn't use format strings, but instead calls a function. It seems like somehow we'd want the function to be able to draw the whole thing differently once the entry field essentially grows here.

**Comment by floam** [comment]
Call it with an argument - number of lines?

**Comment by ridiculousfish** [comment]
I think the case for customizing this prompt is quite weak. Let's approach this as a UI problem: how do we show more input is expected?

**Comment by floam** [comment]
What we do when the user escapes the new line or starts a block is a bit mysterious. A UI problem that exists is it feels really dangerous to hit enter to cause the new line to happen - I'm always pretty scared fish will execute my incomplete command - it usually works I guess.

**Comment by floam** [comment]
I think we should want sone kind if modal indication once you've gotten into 'editing' stuff once that's pretty explicit if enter is 'execute' or 'add a new line' - and not have it vary so much depending on context of the command input from there on out. A cursor?

**Comment by floam** [comment]
Sorry, I missed your comment there @aureooms and made a similar point redundantly - I agree. :)

**Comment by floam** [comment]
Maybe decorate whatever can be identified as the unclosed token with the 'error' style or `rev` capability.

**Comment by floam** [comment]
Or it seems like some along the lines of an ellipsis in the column after the cursor position would make sense.

**Comment by faho** [comment]
>how do we show that more input is expected?

We atually do this for other cases of incomplete input - in particular for unterminated quoted strings and command substitutions.

They'll be highlighted in $fish_color_error. We could do something like this for unterminated blocks - though using that same variable would be confusing as we also use it for non-existent commands.

**Comment by floam** [comment]
I don't like using the error color here - it's not an *error* per-se, it's just in progress.

Here's what I proposed might look like. We're just using `rev` to invert whatever the command was already styled like, meaning nobody will need to customize the color for this:

(run these commands in fish for the "demo" in your theme). The _ is the cursor.

User is defining a function, hitting enter will cause a new line because more input is expected. 
<img width="112" alt="untitled 5" src="https://cloud.githubusercontent.com/assets/291142/18622891/d4dcacce-7dea-11e6-98e9-1fd370055a90.png">

`tput rev; set_color $fish_color_command; echo function (set_color reset) (set_color $fish_color_param)foo(set_color reset)_;`

They closed `function`, hitting enter will execute.
<img width="144" alt="cmmd" src="https://cloud.githubusercontent.com/assets/291142/18622893/db02abb2-7dea-11e6-9f9e-5319bee6e3ae.png">

`set_color $fish_color_command; echo function (set_color reset)(set_color $fish_color_param)foo(set_color reset)\; end\; _;`

**Comment by floam** [comment]
It's not totally right though - I feel like something should happen to the whole line or happen after the cursor. You'd learn what this means, but it's not immediately obvious or intuitive which is different.

**Comment by floam** [comment]
It might make more sense for everything affected from the start of the unclosed thing to the end of input in this case until it is closed, be affected in some manner. Reverse video is a little too much I'd say - italics?

**Comment by ridiculousfish** [comment]
I dislike our current use of `fish_color_error` for incomplete input since it means you don't get syntax feedback until you're done. For example, in a command substitution, you don't get to find out that you mistyped the command until you close the parens.

I strongly dislike reverse video for this, it's much too strong.

We could show something to the right of the cursor or flush right showing the current block, e.g.:

    fish > echo 'hello                          (quote)

maybe a bit longer:

    fish > function foo ;                     (unclosed function)

we could also incorporate this into the prompt, either as a special field (perhaps similar to `fish_mode_prompt`), or via a special variable/command substitution. However this will cause the text to jump around as the indicator text changes.

Probably there are some zsh plugins out there that do this, we should investigate what they do.

**Comment by floam** [comment]
What if we produce the matching bracket, paren, etc. (whatever is necessary to close whatever it is), and place it beyond the cursor, like a text editor might do?

**Comment by ridiculousfish** [comment]
That's a nice idea, as long as it's clear the text is informative and not part of the command line. Inserting it directly into the command line, like many text editors do, would be too far I think.

**Comment by floam** [comment]
Could be an autosuggestion you accept (or not)?

**Comment by floam** [comment]
Although accepting it, you'd want to not actually send the cursor to after it, but keep where it is, so the arrowing would be confusing.

**Comment by krader1961** [comment]
Reclassifying from "question" to "enhancement". But also putting it in the RFC/fish-tank category since it is far from clear we'll reach consensus on how to improve the current situation in the immediate future. In particular it seems like we're definitely not going to introduce a `$PS2` variable so a more fishy solution to the problem is needed.

**Comment by floam** [comment]
Something that doesn't bust syntax highlighting like a faint background color or italics maybe works. Something like what python does can work for multiple line input, but we need to make sure it really does not look like part of the command:

![image](https://user-images.githubusercontent.com/291142/52527336-f7b5d800-2c7b-11e9-9e4b-85190665e324.png)

**Comment by PerBothner** [comment]
If one customizes the background color of the prompt, it make sense to also customize the color of the secondary prompt, even if it is just spaces.

More generally, some terminals support "shell integration", where each command is a structural unit, and where the separate parts of the command (prompt, input, and output) are delimited in the terminal's internal presentation, enabling certain features and extensions.  For this to work reliably it is important to know what is a prompt and what is not.  That can be done by adding escape sequences to delimit the prompt.  We can so that with `fish_prompt` for the main prompt, but we should have a `fish_continuation_prompt` function for continuation lines.  True, a terminal could implement a fish-specific hack by looking for initial sequences of spaces after the initial input line, but that is rather kludgy.

**Comment by PerBothner** [comment]
As a comment to @floam's Sept 19 suggestion:

If I type:

    echo "foo

it seems reasonable to auto-suggest:

    echo "foo|"
where `|` is the cursor.

But what should happen if I type enter?  Should it accept it as a complete command at that point?  Currently a continuation line is created. Perhaps the continuation line should have a `"` suggestion.

**Comment by ridiculousfish** [comment]
I think autosuggesting a closing quote is not a good idea. The reason is small: autosuggestions act as a sort of double-check that what you're typing is valid, and in particular that file paths are valid. If we always suggested a closing quote, there would be no way to distinguish between a valid file path, and an invalid file path with a closing quote.

**Comment by floam** [comment]
Yeah. I can imagine hitting right arrow + enter quickly after seeing the autosuggestion light up and being unhappy with the result (and my impulsivity).

**Comment by melissaboiko** [comment]
I just want a continuation prompt because it's fun to theme.  If my prompt is a little friend waiting for commands, my continuation prompt can be them taking notes:

```fish
╭(｀ω´)╭▸ if test -f ~/file
 .〆(´ω｀)   grep -i pattern ~/file
 .〆(´ω｀) end
A Pattern Language, along with The Timeless Way of Building and
╭(｀ω´)╭▸
```

**Comment by tjapro** [comment]
After read this discussion and the https://github.com/fish-shell/fish-shell/pull/5928 pull request, I write my opinion on this subject:

I'd like to see a **continuation prompt**, nor only because it's fun or hacky, but because the `fish` shell is a enhanced shell, like currently bash, zsh, and PowerShell (as these shells have this feature). Also it can be used with programs like `starship` (https://github.com/starship/starship/pull/3322).

**Comment by AntyElo** [comment]
ghci already has `prompt-cont-function` like PS2, but fish has not. I wish fish has to has something like that.

**Comment by krobelus** [comment]
On Mon, Jun 30, 2025 at 07:39:05PM -0700, anty wrote:
> AntyElo left a comment (fish-shell/fish-shell#3382)
> 
> ghci already has `prompt-cont-function` like PS2, but fish has not. I wish fish has to has something like that.

you can use "if [ (commandline | count) -gt 1 ]; ....; end"

**Comment by AntyElo** [comment]
> you can use "if [ (commandline | count) -gt 1 ]; ....; end"  

But I can not, then something like

```
    if [ (commandline | count) -gt 1 ]
        echo -n '│ '
    else
        echo -n '⟫ '
    end
```

in fish_promt.fish, '│ ' never shown

**Comment by krobelus** [comment]
> But I can not, then something like
> 
> ```
>     if [ (commandline | count) -gt 1 ]
>         echo -n '│ '
>     else
>         echo -n '⟫ '
>     end
> ```
> 
> in fish_promt.fish, '│ ' never shown

only after the next time the prompt is repainted (e.g. ctrl-l).
there's an issue about adding a commandline-changed event,
which you could run "commandline -f repaint" on, I guess.
Though we should find a better solution.

**Comment by injust** [comment]
> What if we produce the matching bracket, paren, etc. (whatever is necessary to close whatever it is), and place it beyond the cursor, like a text editor might do?

That's basically what https://github.com/jorgebucaran/autopair.fish does. Definitely has some edge cases, but it works pretty well for the most part.

> Inserting it directly into the command line, like many text editors do, would be too far I think.

Coming from Warp, that was actually exactly what I wanted.
